### ISSUE TITLE
Bring TUI version to functional parity with GUI version as close as possible

### ISSUE THREAD

**Issue by unxed** [elfmz/far2l#1101]
Some users do not want to leave their preferred terminal apps. For whose making TUI version better will bring greater experience, spreading far2l usage wider. Most of this work is already done, but there are still some minor issues left.

1. Complex hotkeys, like Ctrl+Shift+arrows. Already supported in terminal. Single ESC press — also supported (an option in settings to enable it by default is appreciated).

2. Mouse. Also already works in terminal.

3. Some terminal apps intercept some hotkeys used by far2l. It may be disabled in their settings (instructions on doing this may be added to far2l help optionally).

4. X clipboard access. If far2l extensions are not available, we should fall back to xclip tool usage (AFAIK mc is already doing something like this). There is also `OSC 52` for clipboard access, but it looks like it will never be supported by terminals, so we may postpone its support for now:
https://gitlab.gnome.org/GNOME/vte/-/issues/2495
https://bugs.kde.org/show_bug.cgi?id=372116

5. Desktop notifications. We should fall back to notify-send if far2l extensions are not available

6. Cursor shape. There are escape seqs for modifying it:
https://www.freebsd.org/cgi/man.cgi?query=screen&apropos=0&sektion=4&manpath=FreeBSD+5.4-RELEASE&format=html
E[=s;eC
Set custom cursor shape, where s is the starting and e is the ending scanlines of the cursor.
Other options:
https://unix.stackexchange.com/questions/49485/escape-code-to-change-cursor-shape
https://github.com/kovidgoyal/kitty/issues/715
https://iterm2.com/documentation/2.1/documentation-escape-codes.html

7. BeginConsoleAdhocQuickEdit, GetLargestWindowSize, SetConsoleWindowMaximized — can be done by terminal emulators theirselves.

8. keydown/keyup events — no alfernative for terminal, I afraid. But whose are only used in far2l for F* key suggestions?

9. SetFKeyTitles — no alfernative for terminal, I afraid

**Comment by elfmz** [comment]
It will never be just like, and these small lacks there and here will kill all taste of FAR:) FAR is good not cuz to its FAR, but cuz has many little yet things that makes life easier - spirit of (old good) Windows apps that was lost sometime/sometime :( Just few examples:
1 Complex hotkeys are not distinguishing left and right Ctrl - this is must have to use bookmarks fluently, also not all terminals works same (good/bad) and from my experience no one terminal supports really all possible combinations.
4 Vertical text blocks copy-pasting will be lost (although may think about prefixing text with some magic pattern that would declare clipboard data format).
7 BeginConsoleAdhocQuickEdit is context dependent - if you click on command line or background you've getting auto-copy behaviour, if you're clicking inside editor or panels - you're will get some different functionality.
8/9 you wrong yourself

**Comment by unxed** [comment]
Of course, it's best to keep maximum of the UX of Windows Far, and it makes sense to offer the user the wx version by default in all cases then X is available.

At the same time, we already offer a tty mode, and if this mode spoils the impression of far2l, it will hardly spoil it more with xclip support for example :)

**Comment by Katarn** [comment]
I'm used to using FAR in drop-down, quake-style terminals - both on Windows and on Linux. The WX version cannot be integrated into such terminals in any way. Therefore, I fully support the author of the topic - it would be very good if the headlights for the terminal were as close as possible to the supported functions with the WX version.

**Comment by unxed** [comment]
X interactions in TTY mode (clipboard, keyboard modifiers probing) are complete!
https://github.com/elfmz/far2l/commit/126a52b80cfa47caa52e8dab7bf3d7ee96a53f7b

**Comment by unxed** [comment]
Ой, там протокол extensions опять поменялся, да? А можно кратенькое резюме изменений, пожалуйста — чтобы putty обновить

**Comment by unxed** [comment]
window management (maximization, etc):
https://invisible-island.net/xterm/ctlseqs/ctlseqs.html
look for "Window manipulation" string

**Comment by unxed** [comment]
Ctrl[+Shift]+digits now working in terminals in new ttyxi mode, great! NB, this mode is not suitable for unstalbe connections. For such cases run remote far2l inside local far2l to use far2l terminal extensions for input handling.

**Comment by unxed** [comment]
some usage stats from far2l russian chat (~30 votes total)

![Снимок экрана от 2022-01-02 03-53-57](https://user-images.githubusercontent.com/1151423/147863429-ebb87c28-492d-4db4-aebb-655c86cc3923.png)

**Comment by unxed** [comment]
> keydown/keyup events — no alfernative for terminal, I afraid

Now possible at least for:
iTerm2
kovidgoyal's kitty
Windows Terminal

Details here: #1575

**Comment by unxed** [comment]
> Desktop notifications. We may fall back to notify-send if far2l extensions are not available

Implemented in #2528

**Comment by unxed** [comment]
Although some improvements in this area are still possible, I believe that overall the problem can be considered solved.

**Comment by unxed** [comment]
First post updated. Actual usage stats (🗿 is ~10% "others"):

![Screenshot_20241207_204523_Telegram](https://github.com/user-attachments/assets/2cd7a63f-7c01-4803-9072-ec8dd9673c53)

**Comment by unxed** [comment]
Ubuntu 24.04 LTS has two more terminal emulators with kitty keyboard protocol support:
https://packages.ubuntu.com/noble/alacritty
https://packages.ubuntu.com/noble/foot (Wayland only!)

**Comment by unxed** [comment]
Milestone:

- At least one terminal supporting improved far2l UX can be easily installed on all major platforms.

- For all remaining issues, at least workarounds exist. User is notified about possible workaround he may whish to apply via getting started help topic in proper language (only en & ru until other translations are complete)

- A patch for win32-input-mode support has been implemented for VTE, the basis of default terminals in many IDEs.
